src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs

```diff
@@ -0,0 +1,177 @@
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously waits for this event to be set or for the wait to be canceled.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token used to cancel the wait. If this token is already canceled, this method will first check whether the event is set.</param>
+        public Task WaitAsync(CancellationToken cancellationToken)
+        {
+            Task waitTask = this.WaitAsync();
+            if (waitTask.IsCompleted)
+                return waitTask;
+
+            return WaitAsync(waitTask, cancellationToken);
+        }
+
+        /// <summary>
+        /// Sets the event, atomically completing every task returned by <see cref="O:Nito.AsyncEx.AsyncManualResetEvent.WaitAsync"/>. If the event is already set, this method does nothing.
+        /// </summary>
+        public void Set()
+        {
+            lock (this.mutex)
+            {
+                this.tcs.TrySetResult(null);
+            }
+        }
+
+        /// <summary>
+        /// Resets the event. If the event is already reset, this method does nothing.
+        /// </summary>
+        public void Reset()
+        {
+            lock (this.mutex)
+            {
+                if (this.tcs.Task.IsCompleted)
+                    this.tcs = CreateAsyncTaskSource<object>();
+            }
+        }
+
+        /// <summary>
+        /// Creates a new TCS for use with async code, and which forces its continuations to execute asynchronously.
+        /// </summary>
+        /// <typeparam name="TResult">The type of the result of the TCS.</typeparam>
+        private TaskCompletionSource<TResult> CreateAsyncTaskSource<TResult>()
+        {
+            return new TaskCompletionSource<TResult>(TaskCreationOptions.RunContinuationsAsynchronously);
+        }
+
+        /// <summary>
+        /// Asynchronously waits for the task to complete, or for the cancellation token to be canceled.
+        /// </summary>
+        /// <param name="task">The task to wait for. May not be <c>null</c>.</param>
+        /// <param name="cancellationToken">The cancellation token that cancels the wait.</param>
+        private Task WaitAsync(Task task, CancellationToken cancellationToken)
+        {
+            Guard.NotNull(task, nameof(task));
+
+            if (!cancellationToken.CanBeCanceled)
+                return task;
+
+            if (cancellationToken.IsCancellationRequested)
+                return Task.FromCanceled(cancellationToken);
+
+            return DoWaitAsync(task, cancellationToken);
+        }
+
+        /// <summary>
+        /// Waits for the task to be executed or canceled.
+        /// </summary>
+        /// <remarks>
+        /// Double await construct is used here in order to catch exceptions in child tasks.
+        /// It is similar to await + unwrap but uses less allocations. <see cref="https://stackoverflow.com/questions/34816628/await-await-vs-unwrap"/>
```

Comment:
if you use `<see>` you need to put it into a sentence, or use `seealso` section

so for example

`For more information about double await see <see ...>.`

Reply:
Fixed